torchvision/csrc/io/image/cuda/encode_jpeg_cuda.cpp

```diff
@@ -0,0 +1,232 @@
+    TORCH_CHECK(
+        image.dtype() == torch::kU8, "Input tensor dtype should be uint8");
+
+    TORCH_CHECK(
+        image.device() == device,
+        "All input tensors must be on the same CUDA device when encoding with nvjpeg")
+
+    TORCH_CHECK(
+        image.dim() == 3 && image.numel() > 0,
+        "Input data should be a 3-dimensional tensor");
+
+    TORCH_CHECK(
+        image.size(0) == 3,
+        "The number of channels should be 3, got: ",
+        image.size(0));
+
+    // nvjpeg requires images to be contiguous
+    contig_images.push_back(image.contiguous());
+  }
+
+  // Create global nvJPEG handle
+  std::call_once(::nvjpeg_handle_creation_flag, nvjpeg_init);
+
+  nvjpegEncoderState_t nv_enc_state;
+  nvjpegEncoderParams_t nv_enc_params;
+
+  // initialize nvjpeg structures
+  // these are rather expensive to create and thus will be reused across
+  // multiple calls to encode_single_jpeg
+  try {
+    nvjpegStatus_t stateCreateResult =
+        nvjpegEncoderStateCreate(nvjpeg_handle, &nv_enc_state, stream);
+    TORCH_CHECK(
+        stateCreateResult == NVJPEG_STATUS_SUCCESS,
+        "Failed to create nvjpeg encoder state: ",
+        stateCreateResult);
+
+    nvjpegStatus_t paramsCreateResult =
+        nvjpegEncoderParamsCreate(nvjpeg_handle, &nv_enc_params, stream);
+    TORCH_CHECK(
+        paramsCreateResult == NVJPEG_STATUS_SUCCESS,
+        "Failed to create nvjpeg encoder params: ",
+        paramsCreateResult);
+
+    nvjpegStatus_t paramsQualityStatus =
+        nvjpegEncoderParamsSetQuality(nv_enc_params, quality, stream);
+    TORCH_CHECK(
+        paramsQualityStatus == NVJPEG_STATUS_SUCCESS,
+        "Failed to set nvjpeg encoder params quality: ",
+        paramsQualityStatus);
+
+    std::vector<torch::Tensor> encoded_images;
+    for (const auto& image : contig_images) {
+      auto encoded_image = encode_single_jpeg(
+          image, quality, stream, device, nv_enc_state, nv_enc_params);
+      encoded_images.push_back(encoded_image);
+    }
+    // Clean up
+    nvjpegEncoderStateDestroy(nv_enc_state);
+    nvjpegEncoderParamsDestroy(nv_enc_params);
+    return encoded_images;
+  } catch (const std::exception& e) {
+    nvjpegEncoderStateDestroy(nv_enc_state);
+    nvjpegEncoderParamsDestroy(nv_enc_params);
+    throw;
+  }
+}
+
+torch::Tensor encode_single_jpeg(
+    const torch::Tensor& src_image,
+    const int64_t quality,
+    const cudaStream_t stream,
+    const torch::Device& device,
+    const nvjpegEncoderState_t& nv_enc_state,
+    const nvjpegEncoderParams_t& nv_enc_params) {
+  int channels = src_image.size(0);
+  int height = src_image.size(1);
+  int width = src_image.size(2);
+
+  nvjpegStatus_t samplingSetResult = nvjpegEncoderParamsSetSamplingFactors(
+      nv_enc_params, NVJPEG_CSS_444, stream);
+  TORCH_CHECK(
+      samplingSetResult == NVJPEG_STATUS_SUCCESS,
+      "Failed to set nvjpeg encoder params sampling factors: ",
+      samplingSetResult);
```

Comment:
In https://github.com/pytorch/vision/pull/8375#discussion_r1571016848 you mentioned this may not be needed anymore?

Reply:
It's again one of those things that's not mentioned in the docs, but apparently is mandatory. See here: https://stackoverflow.com/questions/65929613/nvjpeg-encode-packed-bgr